### PR TITLE
Fixes #28065 - package dynflow sidekiq service files

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Depend on:
- [x] Package dependencies #4188
- [x] Prepare init script theforeman/foreman#7109

### CODE REVIEW

File: packages/foreman/foreman/dynflow@.service [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[Unit]
```
Comment (ekohl) [inline packages/foreman/foreman/dynflow@.service]: Can we ship this file in the foreman.git repo? We do this for smart-proxy (https://github.com/theforeman/smart-proxy/blob/develop/extra/systemd) and foreman (https://github.com/theforeman/foreman/tree/develop/extras/systemd) and it has the benefit of being able to share configs between debs and rpms.

File: packages/foreman/foreman/dynflow@.service [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+Description=Foreman jobs daemon on sidekiq
+Documentation=https://theforeman.org
+After=network.target remote-fs.target nss-lookup.target
+
+[Service]
+Type=forking
+User=foreman
+TimeoutSec=90
+ExecStart=/usr/bin/bundle exec sidekiq -e ${RAILS_ENV} -r ${DYNFLOW_SIDEKIQ_SCRIPT} -C /etc/foreman/dynflow/%i.yml
```
Comment (ekohl) [inline packages/foreman/foreman/dynflow@.service]: There is no bundle on production installs.

File: packages/foreman/foreman/dynflow@.service [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+WorkingDirectory=/usr/share/foreman
+EnvironmentFile=-/etc/sysconfig/dynflow
```
Comment (ekohl) [inline packages/foreman/foreman/dynflow@.service]: For new services we don't use `/etc/sysconfig` anymore but prefer systemd drop in configs. This saves us from having to manage the difference between RPMs (/etc/sysconfig) and debs (/etc/default).

File: packages/foreman/foreman/foreman.spec [hunk 4]
```diff
@@ -711,6 +714,38 @@ Meta Package to install requirements for Redis caching support
 %files redis
 %{_datadir}/%{name}/bundler.d/redis.rb
 
+%package dynflow-sidekiq
+Summary: Foreman Redis caching support
+Group:  Applications/System
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec]: Could you add a line to allow automatic updating of this?
```suggestion
Group:  Applications/System
# start specfile sidekiq Requires
```

File: packages/foreman/foreman/foreman.spec [hunk 5]
```diff
@@ -711,6 +714,38 @@ Meta Package to install requirements for Redis caching support
...
+Requires: %{?scl_prefix}rubygem(sidekiq) >= 5.0
+Requires: %{?scl_prefix}rubygem(sidekiq) < 6
+Requires: %{?scl_prefix}rubygem(gitlab-sidekiq-fetcher)
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec]:
```suggestion
Requires: %{?scl_prefix}rubygem(gitlab-sidekiq-fetcher)
# end specfile sidekiq Requires
```

File: packages/foreman/foreman/foreman.spec [hunk 6]
```diff
@@ -2,6 +2,7 @@
 %global confdir extras/packaging/rpm/sources
 %global foreman_rake %{_sbindir}/%{name}-rake
 %global executor_service_name dynflowd
+%global dynflow_sidekiq_service_name dynflow-sidekiq
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec]: I think `@` should be part of the service name

File: packages/foreman/foreman/foreman.spec [hunk 7]
```diff
@@ -711,6 +714,38 @@ Meta Package to install requirements for Redis caching support
...
+%config(noreplace) %{_sysconfdir}/sysconfig/%{dynflow_sidekiq_service_name}
+%{_unitdir}/%{dynflow_sidekiq_service_name}@.service
+%{_datadir}/%{name}/script/dynflow-sidekiq.rb
+
+%install dynflow-sidekiq
+install -d -m0755 %{buildroot}%{_sysconfdir}/%{name}/dynflow
+install -Dp -m0644 %{SOURCE12} %{buildroot}%{_sysconfdir}/sysconfig/%{dynflow_sidekiq_service_name}
+install -Dp -m0644 %{SOURCE13} %{buildroot}%{_unitdir}/%{dynflow_sidekiq_service_name}@.service
+
+sed -i '/^ExecStart/ s|/usr/bin/bundle \(.\+\)$|/usr/bin/scl enable tfm "bundle \1"|' %{buildroot}%{_unitdir}/%{dynflow_sidekiq_service_name}@.service
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec]: This is more of a `%build` step.

File: packages/foreman/foreman/foreman.spec [hunk 8]
```diff
@@ -711,6 +714,38 @@ Meta Package to install requirements for Redis caching support
...
+
+mv %{buildroot}%{_datadir}/%{name}/config/dynflow/orchestrator.yml.example %{buildroot}%{_datadir}/%{name}/config/orchestrator.yml
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec]: Can we move it just once instead of twice?

File: packages/foreman/foreman/foreman.spec [hunk 9]
```diff
@@ -711,6 +713,34 @@ Meta Package to install requirements for Redis caching support
...
+Group:  Applications/System
+# start specfile redis Requires
+Requires: %{?scl_prefix}rubygem(sidekiq) >= 5.0
+Requires: %{?scl_prefix}rubygem(sidekiq) < 6
+Requires: %{?scl_prefix}rubygem(gitlab-sidekiq-fetcher)
+# end specfile redis Requires
+Requires: %{name} = %{version}-%{release}
+
+%description dynflow-sidekiq
+Meta Package to install requirements for Redis caching support
```
Comment (ehelms) [inline packages/foreman/foreman/foreman.spec]: This description mentions Redis caching support but this is for Dynflow support right?

File: packages/foreman/foreman/foreman.spec [hunk 10]
```diff
@@ -711,6 +713,34 @@ Meta Package to install requirements for Redis caching support
 %files redis
 %{_datadir}/%{name}/bundler.d/redis.rb
 
+%package dynflow-sidekiq
```
Comment (ehelms) [inline packages/foreman/foreman/foreman.spec]: Does this package conflict with the regular dynflow service? Can both run at the same time? What happens if they do? Is the new dynflow-sidekiq optional?

File: packages/foreman/foreman/foreman.spec (hunk 9 above)
Comment (ezr-ondrej) [reply packages/foreman/foreman/foreman.spec]: My bad, too much of copy pasting.

File: packages/foreman/foreman/foreman.spec (hunk 10 above)
Comment (ezr-ondrej) [reply packages/foreman/foreman/foreman.spec]: No, it doesn't conflict, and it is not trying to remove it, if someone runs both, they will share the load and loadbalance.
`dynflow-sidekiq` is installed in installer only if the option is turned on (theforeman/puppet-foreman#761) so yes, it's optional.

File: packages/foreman/foreman/foreman.spec [hunk 11]
```diff
@@ -711,6 +713,43 @@ Meta Package to install requirements for Redis caching support
...
+
+%files dynflow-sidekiq
+%{_datadir}/%{name}/bundler.d/sidekiq.rb
+%{_unitdir}/%{dynflow_sidekiq_service_name}.service
+%{_datadir}/%{name}/script/dynflow-sidekiq.rb
+
+%install dynflow-sidekiq
+install -Dp -m0644 extras/systemd/%{dynflow_sidekiq_service_name}.service %{buildroot}%{_unitdir}/%{dynflow_sidekiq_service_name}.service
+
+sed -i '/^ExecStart/ s|/usr/local/bin/sidekiq \(.\+\)$|/usr/bin/scl enable tfm "sidekiq \1"|' %{buildroot}%{_unitdir}/%{dynflow_sidekiq_service_name}.service
```
Comment (adamruzicka) [inline packages/foreman/foreman/foreman.spec]: Does `scl enable tfm "sidekiq -C something"` work? I always used `scl enable tfm -- sidekiq -C something`

Comment (ehelms) [reply packages/foreman/foreman/foreman.spec]: We do a similar thing for running Foreman as a service using puma.

File: packages/foreman/foreman/foreman.spec (hunk 10 above)
Comment (ehelms) [reply packages/foreman/foreman/foreman.spec]: Does the singleton nature of the dynflow sidekiq orchestrator play well with the current dynflow executor if both are deployed simultaneously?

Comment (ezr-ondrej) [reply packages/foreman/foreman/foreman.spec]: Yeah, that doesn't really matter, the separation is not really affecting the old deployment.

File: packages/foreman/foreman/foreman.spec (hunk 11 above)
Comment (ezr-ondrej) [reply packages/foreman/foreman/foreman.spec]: I took it from there actually.

File: packages/foreman/foreman/foreman.spec [hunk 12]
```diff
@@ -27,6 +28,7 @@ Source4: %{name}.cron.d
 Source5: %{name}.tmpfiles
 Source10: %{executor_service_name}.sysconfig
 Source11: %{executor_service_name}.service
+Source12: extras/systemd/%{dynflow_sidekiq_service_name}.service
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec]: This needs to be removed.
```suggestion
```

File: packages/foreman/foreman/foreman.spec [hunk 13]
```diff
@@ -711,6 +713,43 @@ Meta Package to install requirements for Redis caching support
...
+%{_unitdir}/%{dynflow_sidekiq_service_name}.service
+%{_datadir}/%{name}/script/dynflow-sidekiq.rb
+
+%install dynflow-sidekiq
+install -Dp -m0644 extras/systemd/%{dynflow_sidekiq_service_name}.service %{buildroot}%{_unitdir}/%{dynflow_sidekiq_service_name}.service
+
+sed -i '/^ExecStart/ s|/usr/bin/sidekiq \(.\+\)$|/usr/bin/scl enable tfm "sidekiq \1"|' %{buildroot}%{_unitdir}/%{dynflow_sidekiq_service_name}.service
+
+for i in orchestrator.yml worker.yml; do
+mv %{buildroot}%{_datadir}/%{name}/config/dynflow/$i %{buildroot}%{_sysconfdir}/%{name}/dynflow/
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec]: Can you indent these lines?

File: packages/foreman/foreman/foreman.spec [hunk 14]
```diff
@@ -711,6 +713,43 @@ Meta Package to install requirements for Redis caching support
...
+
+%description dynflow-sidekiq
+Meta Package to install dynflow sidekiq executor support
+
+%files dynflow-sidekiq
+%{_datadir}/%{name}/bundler.d/sidekiq.rb
+%{_unitdir}/%{dynflow_sidekiq_service_name}.service
+%{_datadir}/%{name}/script/dynflow-sidekiq.rb
+
+%install dynflow-sidekiq
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec]: I don't think you can have separate `%install` sections. At least, we don't use it anywhere else.

Comment (ezr-ondrej) [reply packages/foreman/foreman/foreman.spec]: So I should install it in the main package directly? or to use build section?

Comment (ekohl) [reply packages/foreman/foreman/foreman.spec]: Main `%install` is fine. Then within `%files` you split it up again.

**Comment by ekohl** [comment]
Also needs an entry in comps.

**Comment by ezr-ondrej** [comment]
Added comps section, moved install to main install.

**Comment by ezr-ondrej** [comment]
Add changelog line

**Comment by ekohl** [comment]
The build failure is due to the nodejs changes that are going on.

**Comment by ehelms** [comment]
This current design installs and starts "classic" dynflow by default. As I understood things, the target was for the new dynflow-sidekiq to be the default?

If we are keeping "classic" dynflow as the default, installing this new subpackage will configure and start dynflow-sidekiq while leaving classic dynflow running as is. Is this what a user should expect? That feels like it will lead to inconsistencies.

**Comment by ezr-ondrej** [comment]
@ehelms this is now addressed in the installer PR theforeman/puppet-foreman#761 it will disable the old service for you once you select the sidekiq to be installed. If you are not using installer you're on your own to figure it out though.

**Comment by tbrisker** [comment]
this is now blocked on the core PR: `BUILDSTDERR: install: cannot stat 'extras/systemd/dynflow-sidekiq@.service': No such file or directory`

**Comment by ezr-ondrej** [comment]
Any idea why the tests are failing now? :disappointed:

**Comment by ekohl** [comment]
Because there wasn't a new source tarball with the new file. There is now.
[test rpm]

File: packages/foreman/foreman/foreman.spec [hunk 15]
```diff
@@ -2,6 +2,7 @@
 %global confdir extras/packaging/rpm/sources
 %global foreman_rake %{_sbindir}/%{name}-rake
 %global executor_service_name dynflowd
+%global dynflow_sidekiq_service_name dynflow-sidekiq@
```
Comment (ekohl) [inline packages/foreman/foreman/foreman.spec]: Given the repetition I'm inclined to add `.service` here but I'm fine either way. I'd probably do a cleanup of all service handling later anyway.

Comment (ezr-ondrej) [reply packages/foreman/foreman/foreman.spec]: I would like to stay consistent with the `dynflowd` service, otherwise I would agree :thinking:

**Comment by ekohl** [comment]
I'm looking at https://github.com/mperham/sidekiq/blob/master/examples/systemd/sidekiq.service and it has some neat tricks:
Reloading:
```
# use `systemctl reload sidekiq` to send the quiet signal to Sidekiq
# at the start of your deploy process.
ExecReload=/usr/bin/kill -TSTP $MAINPID
```

Memory fragmentation fix:
```
# Greatly reduce Ruby memory fragmentation and heap usage
# https://www.mikeperham.com/2018/04/25/taming-rails-memory-bloat/
Environment=MALLOC_ARENA_MAX=2
```

I'm wondering if the `type=forking` that's currently used is incorrect and should be `type=simple`. Does sidekiq fork itself?

**Comment by tbrisker** [comment]
considering that this is experimental in 1.24, perhaps we can merge if it works and improve afterwards if needed?

**Comment by ehelms** [comment]
@tbrisker That was my plan. My comment was to indicate I did not get it to work, and to identify some issues that if we don't fix now, are captured in issues that block it going from "beta" to "GA".

**Comment by ezr-ondrej** [comment]
I am trying to address the comments in theforeman/foreman#7127 - not to forget them.

**Comment by adamruzicka** [comment]
Addressing the start timeout (it indeed was `simple` vs `forking`) and syslog identifier issues in https://github.com/theforeman/foreman/pull/7128

**Comment by tbrisker** [comment]
what's the status here @ehelms @ekohl ?

**Comment by ehelms** [comment]
@tbrisker I wanted to see this in action to know this packaging change plus the code works if someone installs it and then manually configures Redis. I have not reached that point yet, working through it.

**Comment by ekohl** [comment]
Should we add a `%exclude` commit to rpm/develop in the mean time to get nightly to build again?

**Comment by ehelms** [comment]
[test rpm]

**Pull request merged**